### ISSUE TITLE
Create tmp file and rename to dest

### ISSUE THREAD

**Issue by jiqimaogou** [michel-kraemer/gradle-download-task#87]
if I interrupt my download task, the dest is invalidate.

can you use create tmp file first, and rename to dest。

every time to access the network to validate the file is not recommend.

**Comment by michel-kraemer** [comment]
This is already possible. Please have a look at the file `tempRename.gralde` from the `examples` directory:
https://github.com/michel-kraemer/gradle-download-task/blob/master/examples/tempRename.gradle

More information can be found in my blog post:
https://www.michel-kraemer.com/recipes-for-gradle-download/#recipe-8-temporary-part-file-i-e-download-and-rename-

**Comment by shalupov** [comment]
@michel-kraemer I think the original point of this issue is that not having this in core functionality is actually a bug. If you terminate gradle while downloading a file, the plugin actually is not doing what it was intended to do (download a file and cache if for future calls).

From practical perspective we have tens of places with download task, it's not feasible to fix them all.
If course we may wrap it in another task or whatever, but it does not guarantee some fresh developer won't use it directly.

**Comment by michel-kraemer** [comment]
Thanks for the clarification. I understand what you mean. I'm reopening the issue as a reminder to implement this feature in a future version.

Pull requests are welcome, by the way 😄

**Comment by michel-kraemer** [comment]
I added a new flag called `tempAndMove` for this feature. It will be available in the next version.

**Comment by shalupov** [comment]
Thanks!

**Comment by zolotov** [comment]
@michel-kraemer thank you for implementing this. Unfortunately I faced an issue with this on windows, when the existing file is outdated.

To recreate it please follow the steps:

1. create task  with your own `url` and `outputFile` values
```
task myTask {
  doLast {
    download {
        src url
        dest outputFile
        tempAndMove true
      }
  }
}
```

2. run it once
3. modify the output file (e.g. remove it and create an empty file with the same name)
4. run `myTask` one more time

```
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.IllegalStateException: Could not download file
	at de.undercouch.gradle.tasks.download.DownloadExtension.configure(DownloadExtension.java:46)
	at de.undercouch.gradle.tasks.download.DownloadExtension.configure(DownloadExtension.java:29)
	at org.gradle.util.ConfigureUtil.configure(ConfigureUtil.java:104)
	at org.gradle.util.ConfigureUtil$1.execute(ConfigureUtil.java:123)
	at org.gradle.api.internal.plugins.ExtensionsStorage$ExtensionHolder.configure(ExtensionsStorage.java:173)
	at org.gradle.api.internal.plugins.ExtensionsStorage.configureExtension(ExtensionsStorage.java:70)
	at org.gradle.api.internal.plugins.DefaultConvention.configureExtension(DefaultConvention.java:367)
	at org.gradle.api.internal.plugins.DefaultConvention.access$500(DefaultConvention.java:45)
	at org.gradle.api.internal.plugins.DefaultConvention$ExtensionsDynamicObject.tryInvokeMethod(DefaultConvention.java:310)
	at org.gradle.internal.metaobject.CompositeDynamicObject.tryInvokeMethod(CompositeDynamicObject.java:98)
	at org.gradle.internal.metaobject.MixInClosurePropertiesAsMethodsDynamicObject.tryInvokeMethod(MixInClosurePropertiesAsMethodsDynamicObject.java:30)
	at org.gradle.groovy.scripts.BasicScript$ScriptDynamicObject.tryInvokeMethod(BasicScript.java:134)
	at org.gradle.internal.metaobject.ConfigureDelegate.invokeMethod(ConfigureDelegate.java:80)
	at setupJdk_6h1uxoouxz4xk4vhsh5jownip$_run_closure1$_closure13.doCall(C:\Users\g\IdeaProjects\community\build\dependencies\setupJdk.gradle:23)
	at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:722)
	at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:695)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:121)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:110)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
	... 104 more
Caused by: java.io.IOException: Failed to move temporary file 'C:\Users\g\IdeaProjects\community\build\dependencies\build\download-task\jbsdk8u152b1153.5_windows_x64.tar.gz.part' to destination file 'C:\Users\g\IdeaProjects\community\build\dependencies\build\jbsdk\jbsdk8u152b1153.5_windows_x64.tar.gz'.
	at de.undercouch.gradle.tasks.download.DownloadAction.streamAndMove(DownloadAction.java:341)
	at de.undercouch.gradle.tasks.download.DownloadAction.performDownload(DownloadAction.java:314)
	at de.undercouch.gradle.tasks.download.DownloadAction.executeHttpProtocol(DownloadAction.java:273)
	at de.undercouch.gradle.tasks.download.DownloadAction.execute(DownloadAction.java:204)
	at de.undercouch.gradle.tasks.download.DownloadAction.execute(DownloadAction.java:156)
	at de.undercouch.gradle.tasks.download.DownloadExtension.configure(DownloadExtension.java:44)
	... 126 more
```

Could you suggest how to fix it or should I file a separate issue about it?

**Comment by michel-kraemer** [comment]
Thanks for the feedback. Normally, I'd say we should create a separate issue but since we're already here, I'll just reopen this one.

I'm wondering if this is related to the fact that the Gradle daemon keeps file handles open. Can you please kill all gradle processes (or restart your computer) and try to reproduce the error with `gradle --no-daemon`?

**Comment by zolotov** [comment]
@michel-kraemer looks like it's not related. Recreated without a daemon after VM restart.

**Comment by michel-kraemer** [comment]
Thanks for testing. I'll try to get my hands on a Windows machine and test it myself. I'll get back to you as soon as I have some results.

**Comment by shalupov** [comment]
@michel-kraemer I've fixed this, see #110

**Comment by shalupov** [comment]
Also I propose to enable this option by default since others do it and it does not cause problems. I remember rsync and pure-ftpd save files to a temporary location and them rename them.

> From practical perspective we have tens of places with download task, it's not feasible to fix them all.
> If course we may wrap it in another task or whatever, but it does not guarantee some fresh developer won't use it directly.

**Comment by michel-kraemer** [comment]
@shalupov Thanks for the pull request, but I don't think copying the file contents is a good solution. What if you download a very large file (like several GB)? Copying the contents will significantly slow down the whole process.

I just pushed another fix in 16c5d60. This should solve the issue. I just tested it successfully under Windows. I'll publish a new version in a couple of minutes. Please let me know if the problem persists.

**Comment by zolotov** [comment]
3.4.2 works fine for me. Thanks.
